go/runtime/client/watcher.go

```diff
@@ -103,64 +121,112 @@ func (w *blockWatcher) watch() {
 	}
 	defer blocksSub.Close()
 
-	// If we were just started, refresh the committee information from any
-	// block, otherwise just from epoch transition blocks.
-	var gotInitialCommittee bool
+	// Start watching consensus blocks.
+	consensusBlocks, consensusBlocksSub, err := w.common.consensus.WatchBlocks(w.common.ctx)
+	if err != nil {
+		w.Logger.Error("failed to subscribe to consensus blocks",
+			"err", err,
+		)
+		return
+	}
+	defer consensusBlocksSub.Close()
+
+	// latestHeight contains the latest known consensus block height.
+	var latestHeight int64
 	// latestGroupVersion contains the latest known committee group version.
 	var latestGroupVersion int64
+	// Wait for first consensus block before proceeding.
+	select {
+	case <-w.stopCh:
+		return
+	case <-w.common.ctx.Done():
+		return
+	case blk := <-consensusBlocks:
+		latestHeight = blk.Height
+		latestGroupVersion, err = w.getGroupVersion(blk.Height)
+		if err != nil {
+			w.Logger.Error("failed querying for latest group version",
+				"err", err,
+			)
+			return
+		}
+	}
+
 	for {
 		// Wait for stuff to happen.
 		select {
 		case blk := <-blocks:
-			// Check block.
-			w.checkBlock(blk.Block)
+			w.toBeChecked = append(w.toBeChecked, blk.Block)
 
-			// If this is the initial block or an epoch transition block,
-			// update latest known group version and resend all transactions.
-			if gotInitialCommittee && blk.Block.Header.HeaderType != block.EpochTransition {
-				continue
+			var failedBlocks []*block.Block
+			for _, b := range w.toBeChecked {
+				if err = w.checkBlock(b); err != nil {
+					w.Logger.Error("error checking block",
+						"err", err,
+						"block", b,
+					)
+					failedBlocks = append(failedBlocks, b)
+				}
 			}
+			w.toBeChecked = failedBlocks
```

Comment:
Should there be a (configurable) limit on the size of this list? The questions is what should happen once that limit is reached.

Reply:
I'm not sure, if anything i think it might make more sense to have a limit on number of transactions being watched. I guess it wouldn't hurt at least logging a warning in case the `failedBlocks` is non-zero though.

Reply:
But this doesn't have anything to do with the number of watched transactions, right? There could be a single transaction and 1000 blocks waiting to be re-checked? And yeah logging sounds like a good idea.

Reply:
Yeah it's not related. I just think that it's more realistic that the number of transaction will grow big, rather than number of blocks needing to be rechecked, since in case there's actually problem with the storage nodes (the only reason checking blocks can currently fail), there probably won't be any new roothash blocks anyway. Unless the storage nodes are specifically blocking client requests.

Reply:
Oh one thing that would be useful is to clear this list in case there are no more pending transactions? Because it doesn't make sense to re-check blocks if all transactions have either completed or expired.

Reply:
agreed. Hm wondering if instead should just make `checkBlock` first check if there's any pending transactions and in case there aren't any just successfully return. Since it also doesn't make sense checking new blocks in case there aren't any pending transactions.

Reply:
I instead made `checkBlock` check if there's any pending transactions.